### DISCUSSION TITLE
Datamigration without loss

### DISCUSSION THREAD

**Discussion by jkottapa** [prisma/prisma#2193] [Q&A] [answered]
Original schema
```prisma
model Vendor {
  id        Int      @id @default(autoincrement())
  email     String   @unique
  imageUrl String?
  name      String?
  role      Role     @default(VENDOR)
  vendorStrings     VendorString[]
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
}

model VendorString {
  id         Int        @id @default(autoincrement())
  createdAt  DateTime   @default(now())
  updatedAt  DateTime   @updatedAt
  name      String  @unique
  vendor     Vendor      @relation(fields: [vendorId], references: [id])
  vendorId Int
}

enum Role {
  VENDOR
  ADMIN
}
```

I want to make the `vendor` attribute many-to-many relation instead of 1-many relation it is now.
Here is my new schema:
```prisma
model Vendor {
  id        Int      @id @default(autoincrement())
  email     String   @unique
  imageUrl String?
  name      String?
  role      Role     @default(VENDOR)
  vendorStrings     VendorString[] @relation(references: [id])
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
}

model VendorString {
  id         Int        @id @default(autoincrement())
  createdAt  DateTime   @default(now())
  updatedAt  DateTime   @updatedAt
  name      String  @unique
  vendor     Vendor[]      @relation(references: [id])
}

enum Role {
  VENDOR
  ADMIN
}
```

How do I make it happen with `prisma migrate` without losing my existing relations already established?

**Comment by ryands17** [comment, marked as answer]
Hey @jayanth1991 
Currently `prisma migrate` doesn't support directly converting scalar fields to a list without data loss.
For that, you would have to run your migrations in the following manner.
- You would have to create a new array field first
- seed the converted data to the newly created field
- delete the non-required scalar field after all the data is seeded

This is currently the approach that you can follow to safely migrate from 1-n to m-n relations.

**Comment by ryanking1809** [comment]
I just came across this issue with migrating columns like `name` and transforming the data into `firstName` and `lastName`.  Sounds like you already something in mind but would be amazing to be able to execute javascript functions between migration steps in the future! 

Maybe something like this.

```js
steps: [
   {
      "tag": "CreateField",
      "model": "User",
      "field": "firstName",
      "type": "String",
    },
    {
      "tag": "CreateField",
      "model": "User",
      "field": "lastName",
      "type": "String",
    },
    {
      "dataTransformer": splitUserNamesAndSave()
    },
    {
      "tag": "DeleteField",
      "model": "User",
      "field": "name",
    }
]
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
